[TP][Inference] Enable DTensor TP inference

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110751


In https://github.com/pytorch/pytorch/pull/109977, we observed that during inference mode, aten.Linear does not get decomposed. So instead of enabling sharding propagation for linear op, we use func.decompose so that it gets decomposed to matmul and mm.

fduwjj committed:
Enable DTensor TP inference

[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/api.py:
```diff
@@ -253,6 +253,10 @@ def __tensor_unflatten__(inner_tensors, flatten_spec):
     # pyre-fixme[2]: Parameter must be annotated.
     def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        # Decomposes CompositeImplicitAutograd ops
+        r = func.decompose(*args, **kwargs)
```
We can't by default do this, this would bring quite a bit CPU overhead I believe, we should hide this under a flag check. @bdhirsh @Chillee trying to understand this code https://github.com/pytorch/pytorch/blob/main/torch/_ops.py#L539 and I want to see if there's a simple way to check if op need to be decomposed or not, and call this under this flag. Otherwise we have to either implement a linear op rule, or we do special check for `aten.linear.default`

fduwjj replied:
Add a direct op check and synced with @wanchaol offline, we don't want to use an op allowlist here.

wanchaol commented on test/distributed/tensor/parallel/test_tp_examples.py:
```diff
@@ -107,12 +108,38 @@ def _test_mlp_magatron_e2e(self, is_seq_parallel=False, recompute_activation=Fal
         output_tp = model_tp(inp)
         self.assertEqual(output, output_tp)
 
+    def _test_mlp_magatron_inference(self):
```
nit: we should not mentione megatron here, let's just call it sth like test_inference_mode_inference

fduwjj replied:
Removed megatron from test name.

fduwjj committed:
Update on "[TP][Inference] Enable DTensor TP inference"



In https://github.com/pytorch/pytorch/pull/109977, we observed that during inference mode, aten.Linear does not get decomposed. So instead of enabling sharding propagation for linear op, we use func.decompose so that it gets decomposed to matmul and mm.


[ghstack-poisoned]

fduwjj committed:
Update on "[TP][Inference] Enable DTensor TP inference"



In https://github.com/pytorch/pytorch/pull/109977, we observed that during inference mode, aten.Linear does not get decomposed. So instead of enabling sharding propagation for linear op, we use func.decompose so that it gets decomposed to matmul and mm.


[ghstack-poisoned]

fduwjj committed:
Update on "[TP][Inference] Enable DTensor TP inference"



In https://github.com/pytorch/pytorch/pull/109977, we observed that during inference mode, aten.Linear does not get decomposed. So instead of enabling sharding propagation for linear op, we use func.decompose so that it gets decomposed to matmul and mm.


[ghstack-poisoned]

bdhirsh commented on torch/distributed/_tensor/api.py:
```diff
@@ -253,6 +254,14 @@ def __tensor_unflatten__(inner_tensors, flatten_spec):
+        # This is mostly for inference mode when we want to
+        # decompose CompositeImplicitAutograd ops.
+        # For the long run, we need to think of a better way to handle it.
+        if func == aten.linear.default:
```
You actually can just remove this check for linear - `func.decompose()` will work for **all** CompositeImplicitAutograd ops, and properly return a `NotImplemented` object if it sees that the op has no `CompositeImplicitAutograd` implementation.

If you're worried about hurting eager perf - maybe it's worth a benchmark, but I think the cost of the checks that `decompose()` performs is pretty cheap.

wanchaol replied:
Yeah we should do some benchmarking, @fduwjj let's explicitly add a TODO item here, we can probably land this first and test out decompose() as follow up

fduwjj replied:
sure, let me add a TODO here and do some benchmark and then we can see if we can remove that check.

bdhirsh reviewed (approved):
left light comments but lgtm!

fduwjj committed:
Update on "[TP][Inference] Enable DTensor TP inference"



In https://github.com/pytorch/pytorch/pull/109977, we observed that during inference mode, aten.Linear does not get decomposed. So instead of enabling sharding propagation for linear op, we use func.decompose so that it gets decomposed to matmul and mm.


[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge

fduwjj committed:
Update on "[TP][Inference] Enable DTensor TP inference"



In https://github.com/pytorch/pytorch/pull/109977, we observed that during inference mode, aten.Linear does not get decomposed. So instead of enabling sharding propagation for linear op, we use func.decompose so that it gets decomposed to matmul and mm.


[ghstack-poisoned]

fduwjj committed:
Update on "[TP][Inference] Enable DTensor TP inference"



In https://github.com/pytorch/pytorch/pull/109977, we observed that during inference mode, aten.Linear does not get decomposed. So instead of enabling sharding propagation for linear op, we use func.decompose so that it gets decomposed to matmul and mm.


[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge -f "The failing test are not related to this PR."
